File: source/m-files/loadNikonData.m
function options = loadNikonData(options)
%UNTITLED Summary of this function goes here
%   Detailed explanation goes here
if ~isfield(options, 'fpath') || isempty(options.fpath)
    [file, fpath] = uigetfile('*.xtekct','Select Nikon xtekct file');
    if isequal(file, 0)
        error('No file was selected')
    end
    nimi = [fpath file];
    [fpath,file,ext] = fileparts(nimi);
else
    if exist(options.fpath, 'file') == 0
        warning(['Input file ' options.fpath ' not found!'])
        [file, fpath] = uigetfile('*.xtekct','Select Nikon xtekct file');
        if isequal(file, 0)
            error('No file was selected')
        end
        nimi = [fpath file];
        [fpath,file,ext] = fileparts(nimi);
    else
        [fpath,file,ext] = fileparts(options.fpath);
        nimi = [fpath filesep file ext];
    end
    % nimi = options.fpath;
    % fpath = nimi(1:end-7);
end
fid = fopen(nimi);
hdr = textscan(fid,'%s','Delimiter','=');
hdr = hdr{1};
fclose(fid);
if isfield(options, 'binning') 
    binning = options.binning;
else
    binning = 1;
end
if isfield(options, 'only_reconstructions') 
    only_reconstructions = options.only_reconstructions;
else
    only_reconstructions = false;
end
if ~isfield(options,'Nx')
    ind = find(strcmp(hdr,'VoxelsX')) + 1;
    options.Nx = str2double(hdr{ind(1)});
end
if ~isfield(options,'Ny')
    ind = find(strcmp(hdr,'VoxelsY')) + 1;
    options.Ny = str2double(hdr{ind(1)});
end
if ~isfield(options,'Nz')
    ind = find(strcmp(hdr,'VoxelsZ')) + 1;
    options.Nz = str2double(hdr{ind(1)});
end
ind = find(strcmp(hdr,'DetectorPixelsY')) + 1;
options.nRowsD = str2double(hdr{ind(1)}) / binning;
ind = find(strcmp(hdr,'DetectorPixelsX')) + 1;
options.nColsD = str2double(hdr{ind(1)}) / binning;
ind = find(strcmp(hdr,'DetectorPixelSizeY')) + 1;
options.dPitchX = str2double(hdr{ind(1)}) * binning;
ind = find(strcmp(hdr,'DetectorPixelSizeX')) + 1;
options.dPitchY = str2double(hdr{ind(1)}) * binning;
ind = find(strcmp(hdr,'SrcToDetector')) + 1;
options.sourceToDetector = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'SrcToObject')) + 1;
options.sourceToCRot = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'ObjectOffsetX')) + 1;
options.oOffsetX = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'ObjectOffsetY')) + 1;
options.oOffsetY = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'DetectorOffsetY')) + 1;
options.detOffsetRow = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'DetectorOffsetX')) + 1;
options.detOffsetCol = str2double(hdr{ind(1)});
ind = find(strcmp(hdr,'MaskRadius')) + 1;
options.MaskRadius = str2double(hdr{ind(1)});
fileID = fopen([fpath filesep file '.ang']);
data = textscan(fileID,'%s');
options.nProjections = (numel(data{1}) - 2) / 2;
fclose(fileID);
options.angles = zeros(options.nProjections,1);
data = data{1}(3:end);
for kk = 1 : options.nProjections
    options.angles(kk) = -str2double(data{kk * 2});
end
if ~only_reconstructions || ~isfield(options,'SinM')
    fpath = [fpath filesep file '_0001.tif'];
    options.SinM = loadProjectionImages(options.nProjections,options.binning, fpath);
    options.SinM = permute(options.SinM, [2 1 3]);
end
end

File: source/Python/omegatomo/fileio/loadNikonData.py
# -*- coding: utf-8 -*-


def loadNikonData(options):
    import numpy as np
    import os
    import tkinter as tk
    from tkinter.filedialog import askopenfilename
    from omegatomo.fileio.loadProjectionImages import loadProjectionImages
    def select_file():
    
        if len(options.fpath) == 0:
            root = tk.Tk()
            root.withdraw()
            nimi = askopenfilename(title='Select Nikon xtekct file',filetypes=([('xtekct Files','*.xtekct')]))
            if not nimi:
                raise ValueError('No file was selected')
        else:
            if not os.path.exists(options.fpath):
                print('Specified file was not found! Please select Nikon xtekct file')
                root = tk.Tk()
                root.withdraw()
                nimi = askopenfilename(title='Select Nikon xtekct file',filetypes=([('xtekct Files','*.xtekct')]))
                if not nimi:
                    raise ValueError('No file was selected')
            else:
                nimi = options.fpath
        
        file = os.path.split(nimi)[1]
        file, ext = os.path.splitext(file)
        fpath = os.path.split(nimi)[0] + '/'
        return file, fpath, nimi
    
    file, fpath, nimi = select_file()
    
    
    with open(nimi, 'r') as fid:
        hdr = [line.strip() for line in fid.readlines()]

    if not hasattr(options, 'binning'):
        options.binning = 1
    if not hasattr(options, 'only_reconstructions'):
        options.only_reconstructions = False


    for i in range(len(hdr)):
        if not hasattr(options, 'Nx'):
            if hdr[i].split("=")[0] == 'VoxelsX':
                options.Nx = int(hdr[i].split("=")[1])
        if not hasattr(options, 'Ny'):
            if hdr[i].split("=")[0] == 'VoxelsY':
                options.Ny = int(hdr[i].split("=")[1])
        if not hasattr(options, 'Nz'):
            if hdr[i].split("=")[0] == 'VoxelsZ':
                options.Nz = int(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'DetectorPixelsY':
            options.nRowsD = int(hdr[i].split("=")[1]) // options.binning
        if hdr[i].split("=")[0] == 'DetectorPixelsX':
            options.nColsD = int(hdr[i].split("=")[1]) // options.binning
        if hdr[i].split("=")[0] == 'DetectorPixelSizeY':
            options.dPitchX = float(hdr[i].split("=")[1]) * options.binning
        if hdr[i].split("=")[0] == 'DetectorPixelSizeX':
            options.dPitchY = float(hdr[i].split("=")[1]) * options.binning
        if hdr[i].split("=")[0] == 'SrcToDetector':
            options.sourceToDetector = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'SrcToObject':
            options.sourceToCRot = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'ObjectOffsetX':
            options.oOffsetX = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'ObjectOffsetY':
            options.oOffsetY = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'DetectorOffsetY':
            options.detOffsetRow = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'DetectorOffsetX':
            options.detOffsetCol = float(hdr[i].split("=")[1])
        if hdr[i].split("=")[0] == 'MaskRadius':
            options.MaskRadius = float(hdr[i].split("=")[1])

    with open(os.path.join(fpath, file + '.ang'), 'r') as fid:
        data = [line.strip() for line in fid.readlines()]
    options.nProjections = (len(data) - 1)
    options.angles = np.zeros(options.nProjections, dtype=np.float32)
    data = data[1:]
    for kk in range(options.nProjections):
        parts = data[kk].split(":")
        options.angles[kk] = -float(parts[1].strip())

    if not hasattr(options, 'only_reconstructions') or not options.only_reconstructions:
        file_path = os.path.join(fpath, file + '_0001.tif')
        options.SinM = loadProjectionImages(options.nProjections, options.binning, file_path)
        options.SinM = np.transpose(options.SinM, (1, 0, 2))